addons/mrp/models/product.py

```diff
@@ -186,20 +187,27 @@ def _compute_quantities_dict(self, lot_id, owner_id, package_id, from_date=False
         # compute kit quantities
         for product in bom_kits:
             bom_sub_lines = bom_sub_lines_per_kit[product]
+            # group lines by component
+            bom_sub_lines_grouped = collections.defaultdict(list)
+            for info in bom_sub_lines:
+                bom_sub_lines_grouped[info[0].product_id].append(info)
             ratios_virtual_available = []
             ratios_qty_available = []
             ratios_incoming_qty = []
             ratios_outgoing_qty = []
             ratios_free_qty = []
-            for bom_line, bom_line_data in bom_sub_lines:
-                component = bom_line.product_id.with_context(mrp_compute_quantities=qties).with_prefetch(prefetch_component_ids)
-                if component.type != 'product' or float_is_zero(bom_line_data['qty'], precision_rounding=bom_line.product_uom_id.rounding):
-                    # As BoMs allow components with 0 qty, a.k.a. optionnal components, we simply skip those
-                    # to avoid a division by zero. The same logic is applied to non-storable products as those
-                    # products have 0 qty available.
-                    continue
-                uom_qty_per_kit = bom_line_data['qty'] / bom_line_data['original_qty']
-                qty_per_kit = bom_line.product_uom_id._compute_quantity(uom_qty_per_kit, bom_line.product_id.uom_id, round=False, raise_if_failure=False)
+
+            for component, bom_sub_lines in bom_sub_lines_grouped.items():
+                component = component.with_context(mrp_compute_quantities=qties).with_prefetch(prefetch_component_ids)
+                qty_per_kit = 0
+                for bom_line, bom_line_data in bom_sub_lines:
+                    if component.type != 'product' or float_is_zero(bom_line_data['qty'], precision_rounding=bom_line.product_uom_id.rounding):
+                        # As BoMs allow components with 0 qty, a.k.a. optionnal components, we simply skip those
+                        # to avoid a division by zero. The same logic is applied to non-storable products as those
+                        # products have 0 qty available.
+                        continue
+                    uom_qty_per_kit = bom_line_data['qty'] / bom_line_data['original_qty']
+                    qty_per_kit += bom_line.product_uom_id._compute_quantity(uom_qty_per_kit, bom_line.product_id.uom_id, round=False, raise_if_failure=False)
```

Comment:
I think you can only keep this line

Reply:
What do you mean? :)

Reply:
When testing the same flow locally, I saw replacing `=` by `+=` in this line (and initializing the variable to 0 of course) fixed the issue as well

Reply:
I don't think we can :) See the new version of the test I pushed (note P4 component), it fails if we just add them. We could do as you suggest if we group the `qty_per_kit` per component in a dict for example. But it would be equivalent to what I'm doing. I opted by the version of pre-grouping components because IMHO it makes the computation clearer: we are computing the `qty_per_kit` per component once, and we insert that value in `ratios_virtual_available` only once also.

Let me know if you prefer to make `qty_per_kit` a dict with key `component`.

Reply:
ok ok I see. you can keep it like this 
to me it's ok to merge if you want to make your PR ready to review